File: frontend/src/types/sdk.ts
/**
 * SDK & API Management Types
 *
 * TypeScript types for API keys, usage tracking, rate limits, and code generation.
 */

// Enums

export enum APIKeyType {
  PUBLIC = 'public',
  PRIVATE = 'private',
  RESTRICTED = 'restricted',
}

export enum APIKeyStatus {
  ACTIVE = 'active',
  REVOKED = 'revoked',
  EXPIRED = 'expired',
  SUSPENDED = 'suspended',
}

export enum SDKLanguage {
  JAVASCRIPT = 'javascript',
  TYPESCRIPT = 'typescript',
  PYTHON = 'python',
  REACT = 'react',
  VUE = 'vue',
  ANGULAR = 'angular',
  CURL = 'curl',
  GO = 'go',
  RUBY = 'ruby',
  PHP = 'php',
}

export enum EndpointCategory {
  AUTH = 'auth',
  CONNECTIONS = 'connections',
  DASHBOARDS = 'dashboards',
  CHARTS = 'charts',
  QUERIES = 'queries',
  DATASETS = 'datasets',
  TRANSFORMS = 'transforms',
  MODELS = 'models',
  KPI = 'kpi',
  AI = 'ai',
  ADMIN = 'admin',
}

export enum RateLimitPeriod {
  SECOND = 'second',
  MINUTE = 'minute',
  HOUR = 'hour',
  DAY = 'day',
  MONTH = 'month',
}

// API Key Types

export interface APIKeyPermissions {
  can_read_dashboards: boolean;
  can_write_dashboards: boolean;
  can_read_charts: boolean;
  can_write_charts: boolean;
  can_read_connections: boolean;
  can_write_connections: boolean;
  can_execute_queries: boolean;
  can_read_datasets: boolean;
  can_write_datasets: boolean;
  can_read_transforms: boolean;
  can_write_transforms: boolean;
  can_read_models: boolean;
  can_write_models: boolean;
  can_use_ai: boolean;
  can_embed: boolean;
  can_export: boolean;
  can_admin: boolean;
  workspace_ids: string[];
  dashboard_ids: string[];
}

export interface RateLimitConfig {
  requests_per_second?: number | null;
  requests_per_minute?: number | null;
  requests_per_hour?: number | null;
  requests_per_day?: number | null;
  requests_per_month?: number | null;
  burst_limit: number;
  enabled: boolean;
}

export interface APIKey {
  id: string;
  name: string;
  description?: string | null;
  key_prefix: string;
  key_hash: string;
  key_type: APIKeyType;
  status: APIKeyStatus;
  user_id: string;
  workspace_id?: string | null;
  organization_id?: string | null;
  permissions: APIKeyPermissions;
  rate_limits: RateLimitConfig;
  allowed_ips: string[];
  allowed_domains: string[];
  allowed_referrers: string[];
  expires_at?: string | null;
  last_used_at?: string | null;
  created_at: string;
  updated_at?: string | null;
  created_by: string;
  tags: string[];
  metadata: Record<string, unknown>;
}

export interface APIKeyCreate {
  name: string;
  description?: string | null;
  key_type?: APIKeyType;
  workspace_id?: string | null;
  permissions?: Partial<APIKeyPermissions> | null;
  rate_limits?: Partial<RateLimitConfig> | null;
  allowed_ips?: string[];
  allowed_domains?: string[];
  expires_in_days?: number | null;
  tags?: string[];
}

export interface APIKeyUpdate {
  name?: string | null;
  description?: string | null;
  permissions?: Partial<APIKeyPermissions> | null;
  rate_limits?: Partial<RateLimitConfig> | null;
  allowed_ips?: string[] | null;
  allowed_domains?: string[] | null;
  status?: APIKeyStatus | null;
  tags?: string[] | null;
}

export interface APIKeyResponse {
  api_key: APIKey;
  key?: string | null; // Full key only on creation
}

export interface APIKeyListResponse {
  keys: APIKey[];
  total: number;
}

// Usage Tracking Types

export interface UsageRecord {
  id: string;
  api_key_id: string;
  endpoint: string;
  method: string;
  status_code: number;
  response_time_ms: number;
  request_size_bytes: number;
  response_size_bytes: number;
  ip_address?: string | null;
  user_agent?: string | null;
  timestamp: string;
  metadata: Record<string, unknown>;
}

export interface UsageStats {
  api_key_id: string;
  period_start: string;
  period_end: string;
  total_requests: number;
  successful_requests: number;
  failed_requests: number;
  rate_limited_requests: number;
  requests_by_endpoint: Record<string, number>;
  requests_by_method: Record<string, number>;
  requests_by_status: Record<number, number>;
  avg_response_time_ms: number;
  p50_response_time_ms: number;
  p95_response_time_ms: number;
  p99_response_time_ms: number;
  total_request_bytes: number;
  total_response_bytes: number;
  unique_ips: number;
  unique_endpoints: number;
}

export interface UsageStatsResponse {
  stats: UsageStats;
  daily_breakdown: Array<{
    date: string;
    requests: number;
    errors: number;
  }>;
  top_endpoints: Array<{
    endpoint: string;
    count: number;
  }>;
}

// Rate Limit Types

export interface RateLimitStatus {
  api_key_id: string;
  period: RateLimitPeriod;
  limit: number;
  remaining: number;
  reset_at: string;
  is_limited: boolean;
}

// Code Generation Types

export interface CodeSnippet {
  language: SDKLanguage;
  code: string;
  filename?: string | null;
  description?: string | null;
  dependencies: string[];
}

export interface CodeGenerationRequest {
  language: SDKLanguage;
  resource_type: string;
  resource_id: string;
  include_auth?: boolean;
  include_error_handling?: boolean;
  include_types?: boolean;
  options?: Record<string, unknown>;
}

export interface CodeGenerationResponse {
  snippets: CodeSnippet[];
  setup_instructions: string;
  documentation_url?: string | null;
}

// API Documentation Types

export interface APIParameter {
  name: string;
  type: string;
  required: boolean;
  description?: string | null;
  default?: unknown;
  enum_values: string[];
  example?: unknown;
}

export interface APIEndpoint {
  path: string;
  method: string;
  summary: string;
  description?: string | null;
  category: EndpointCategory;
  tags: string[];
  path_params: APIParameter[];
  query_params: APIParameter[];
  body_schema?: Record<string, unknown> | null;
  response_schema?: Record<string, unknown> | null;
  response_examples: Record<string, unknown>;
  requires_auth: boolean;
  required_permissions: string[];
  rate_limited: boolean;
  deprecated: boolean;
  version: string;
}

export interface APIEndpointGroup {
  name: string;
  description?: string | null;
  category: EndpointCategory;
  endpoints: APIEndpoint[];
}

export interface APIDocumentation {
  title: string;
  version: string;
  description: string;
  base_url: string;
  auth_description: string;
  auth_examples: Record<string, string>;
  groups: APIEndpointGroup[];
  schemas: Record<string, unknown>;
  sdk_languages: SDKLanguage[];
  sdk_download_urls: Record<string, string>;
}

// SDK Configuration Types

export interface SDKConfig {
  version: string;
  base_url: string;
  embed_url: string;
  cdn_url: string;
  supported_languages: string[];
  features: {
    embed: boolean;
    export: boolean;
    interactions: boolean;
    comments: boolean;
    ai: boolean;
  };
}

export interface EmbedInfo {
  embed_url: string;
  iframe_code: string;
  js_snippet: string;
}

// Constants

export const API_KEY_TYPE_LABELS: Record<APIKeyType, string> = {
  [APIKeyType.PUBLIC]: 'Public',
  [APIKeyType.PRIVATE]: 'Private',
  [APIKeyType.RESTRICTED]: 'Restricted',
};

export const API_KEY_STATUS_LABELS: Record<APIKeyStatus, string> = {
  [APIKeyStatus.ACTIVE]: 'Active',
  [APIKeyStatus.REVOKED]: 'Revoked',
  [APIKeyStatus.EXPIRED]: 'Expired',
  [APIKeyStatus.SUSPENDED]: 'Suspended',
};

export const API_KEY_STATUS_COLORS: Record<APIKeyStatus, string> = {
  [APIKeyStatus.ACTIVE]: 'green',
  [APIKeyStatus.REVOKED]: 'red',
  [APIKeyStatus.EXPIRED]: 'gray',
  [APIKeyStatus.SUSPENDED]: 'yellow',
};

export const SDK_LANGUAGE_INFO: Record<SDKLanguage, { name: string; icon: string }> = {
  [SDKLanguage.JAVASCRIPT]: { name: 'JavaScript', icon: 'file-code' },
  [SDKLanguage.TYPESCRIPT]: { name: 'TypeScript', icon: 'file-code' },
  [SDKLanguage.PYTHON]: { name: 'Python', icon: 'file-code' },
  [SDKLanguage.REACT]: { name: 'React', icon: 'atom' },
  [SDKLanguage.VUE]: { name: 'Vue', icon: 'file-code' },
  [SDKLanguage.ANGULAR]: { name: 'Angular', icon: 'file-code' },
  [SDKLanguage.CURL]: { name: 'cURL', icon: 'terminal' },
  [SDKLanguage.GO]: { name: 'Go', icon: 'file-code' },
  [SDKLanguage.RUBY]: { name: 'Ruby', icon: 'gem' },
  [SDKLanguage.PHP]: { name: 'PHP', icon: 'file-code' },
};

export const RATE_LIMIT_PERIOD_LABELS: Record<RateLimitPeriod, string> = {
  [RateLimitPeriod.SECOND]: 'Per Second',
  [RateLimitPeriod.MINUTE]: 'Per Minute',
  [RateLimitPeriod.HOUR]: 'Per Hour',
  [RateLimitPeriod.DAY]: 'Per Day',
  [RateLimitPeriod.MONTH]: 'Per Month',
};

// Helper Functions

export function maskAPIKey(key: string): string {
  if (key.length < 12) return '*'.repeat(key.length);
  return `${key.slice(0, 8)}...${key.slice(-4)}`;
}

export function formatKeyPrefix(prefix: string): string {
  return prefix.replace('bv_', 'bv_****_');
}

export function isKeyExpired(key: APIKey): boolean {
  if (!key.expires_at) return false;
  return new Date(key.expires_at) < new Date();
}

export function getKeyStatusColor(key: APIKey): string {
  if (key.status !== APIKeyStatus.ACTIVE) {
    return API_KEY_STATUS_COLORS[key.status];
  }
  if (isKeyExpired(key)) {
    return 'gray';
  }
  return 'green';
}

export function formatBytes(bytes: number): string {
  if (bytes === 0) return '0 B';
  const k = 1024;
  const sizes = ['B', 'KB', 'MB', 'GB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return `${parseFloat((bytes / Math.pow(k, i)).toFixed(2))} ${sizes[i]}`;
}

export function formatNumber(num: number): string {
  if (num >= 1000000) return `${(num / 1000000).toFixed(1)}M`;
  if (num >= 1000) return `${(num / 1000).toFixed(1)}K`;
  return num.toString();
}

export function formatResponseTime(ms: number): string {
  if (ms < 1000) return `${ms}ms`;
  return `${(ms / 1000).toFixed(2)}s`;
}

export function calculateUsagePercentage(used: number, limit: number): number {
  if (limit === 0) return 0;
  return Math.min(100, (used / limit) * 100);
}

export function getUsageColor(percentage: number): string {
  if (percentage >= 90) return 'red';
  if (percentage >= 75) return 'yellow';
  return 'green';
}

// Default Values

export function createDefaultPermissions(): APIKeyPermissions {
  return {
    can_read_dashboards: true,
    can_write_dashboards: false,
    can_read_charts: true,
    can_write_charts: false,
    can_read_connections: true,
    can_write_connections: false,
    can_execute_queries: true,
    can_read_datasets: true,
    can_write_datasets: false,
    can_read_transforms: true,
    can_write_transforms: false,
    can_read_models: true,
    can_write_models: false,
    can_use_ai: false,
    can_embed: true,
    can_export: true,
    can_admin: false,
    workspace_ids: [],
    dashboard_ids: [],
  };
}

export function createDefaultRateLimits(): RateLimitConfig {
  return {
    requests_per_minute: 60,
    requests_per_hour: 1000,
    requests_per_day: 10000,
    burst_limit: 10,
    enabled: true,
  };
}

// Copy to Clipboard

export async function copyToClipboard(text: string): Promise<boolean> {
  try {
    await navigator.clipboard.writeText(text);
    return true;
  } catch {
    // Fallback for older browsers
    const textarea = document.createElement('textarea');
    textarea.value = text;
    document.body.appendChild(textarea);
    textarea.select();
    const success = document.execCommand('copy');
    document.body.removeChild(textarea);
    return success;
  }
}

// Code Highlighting (simple version)

export function highlightCode(code: string, language: SDKLanguage): string {
  // Basic keyword highlighting for display
  const keywords: Record<string, string[]> = {
    javascript: ['const', 'let', 'var', 'function', 'async', 'await', 'import', 'from', 'export', 'return', 'if', 'else', 'for', 'while'],
    typescript: ['const', 'let', 'var', 'function', 'async', 'await', 'import', 'from', 'export', 'return', 'if', 'else', 'for', 'while', 'interface', 'type'],
    python: ['def', 'class', 'import', 'from', 'return', 'if', 'else', 'for', 'while', 'with', 'as', 'async', 'await'],
  };

  const langKeywords = keywords[language] || [];

  let highlighted = code;
  langKeywords.forEach((keyword) => {
    const regex = new RegExp(`\\b${keyword}\\b`, 'g');
    highlighted = highlighted.replace(regex, `<span class="keyword">${keyword}</span>`);
  });

  // Highlight strings
  highlighted = highlighted.replace(/(["'`])([^"'`]*)\1/g, '<span class="string">$&</span>');

  // Highlight comments
  highlighted = highlighted.replace(/(\/\/.*$|#.*$)/gm, '<span class="comment">$&</span>');

  return highlighted;
}

// API Key Form State

export interface APIKeyFormState {
  name: string;
  description: string;
  keyType: APIKeyType;
  workspaceId: string;
  permissions: APIKeyPermissions;
  rateLimits: RateLimitConfig;
  allowedIps: string[];
  allowedDomains: string[];
  expiresInDays: number | null;
  tags: string[];
}

export function createInitialFormState(): APIKeyFormState {
  return {
    name: '',
    description: '',
    keyType: APIKeyType.PRIVATE,
    workspaceId: '',
    permissions: createDefaultPermissions(),
    rateLimits: createDefaultRateLimits(),
    allowedIps: [],
    allowedDomains: [],
    expiresInDays: null,
    tags: [],
  };
}

export function formStateToCreateRequest(state: APIKeyFormState): APIKeyCreate {
  return {
    name: state.name,
    description: state.description || null,
    key_type: state.keyType,
    workspace_id: state.workspaceId || null,
    permissions: state.permissions,
    rate_limits: state.rateLimits,
    allowed_ips: state.allowedIps,
    allowed_domains: state.allowedDomains,
    expires_in_days: state.expiresInDays,
    tags: state.tags,
  };
}

File: backend/app/schemas/sdk.py
"""
SDK & API Documentation Schemas

Pydantic schemas for SDK tokens, API keys, rate limiting, and code generation.
"""

from typing import Any, Optional
from pydantic import BaseModel, Field
from enum import Enum
from datetime import datetime


# Enums

class APIKeyType(str, Enum):
    """Types of API keys"""
    PUBLIC = "public"  # Read-only, client-side safe
    PRIVATE = "private"  # Full access, server-side only
    RESTRICTED = "restricted"  # Custom permissions


class APIKeyStatus(str, Enum):
    """API key status"""
    ACTIVE = "active"
    REVOKED = "revoked"
    EXPIRED = "expired"
    SUSPENDED = "suspended"


class SDKLanguage(str, Enum):
    """Supported SDK languages"""
    JAVASCRIPT = "javascript"
    TYPESCRIPT = "typescript"
    PYTHON = "python"
    REACT = "react"
    VUE = "vue"
    ANGULAR = "angular"
    CURL = "curl"
    GO = "go"
    RUBY = "ruby"
    PHP = "php"


class EndpointCategory(str, Enum):
    """API endpoint categories"""
    AUTH = "auth"
    CONNECTIONS = "connections"
    DASHBOARDS = "dashboards"
    CHARTS = "charts"
    QUERIES = "queries"
    DATASETS = "datasets"
    TRANSFORMS = "transforms"
    MODELS = "models"
    KPI = "kpi"
    AI = "ai"
    ADMIN = "admin"


class RateLimitPeriod(str, Enum):
    """Rate limit time periods"""
    SECOND = "second"
    MINUTE = "minute"
    HOUR = "hour"
    DAY = "day"
    MONTH = "month"


# API Key Models

class APIKeyPermissions(BaseModel):
    """Permissions for an API key"""
    # Resource permissions
    can_read_dashboards: bool = True
    can_write_dashboards: bool = False
    can_read_charts: bool = True
    can_write_charts: bool = False
    can_read_connections: bool = True
    can_write_connections: bool = False
    can_execute_queries: bool = True
    can_read_datasets: bool = True
    can_write_datasets: bool = False
    can_read_transforms: bool = True
    can_write_transforms: bool = False
    can_read_models: bool = True
    can_write_models: bool = False

    # Advanced permissions
    can_use_ai: bool = False
    can_embed: bool = True
    can_export: bool = True
    can_admin: bool = False

    # Workspace restrictions
    workspace_ids: list[str] = Field(default_factory=list)  # Empty = all workspaces
    dashboard_ids: list[str] = Field(default_factory=list)  # Empty = all dashboards


class RateLimitConfig(BaseModel):
    """Rate limiting configuration"""
    requests_per_second: Optional[int] = None
    requests_per_minute: Optional[int] = 60
    requests_per_hour: Optional[int] = 1000
    requests_per_day: Optional[int] = 10000
    requests_per_month: Optional[int] = None
    burst_limit: int = 10  # Max burst above rate limit
    enabled: bool = True


class APIKey(BaseModel):
    """API key model"""
    id: str
    name: str
    description: Optional[str] = None
    key_prefix: str  # First 8 chars for identification
    key_hash: str  # Hashed full key

    key_type: APIKeyType = APIKeyType.PRIVATE
    status: APIKeyStatus = APIKeyStatus.ACTIVE

    # Ownership
    user_id: str
    workspace_id: Optional[str] = None
    organization_id: Optional[str] = None

    # Permissions
    permissions: APIKeyPermissions = Field(default_factory=APIKeyPermissions)
    rate_limits: RateLimitConfig = Field(default_factory=RateLimitConfig)

    # Security
    allowed_ips: list[str] = Field(default_factory=list)  # Empty = all IPs
    allowed_domains: list[str] = Field(default_factory=list)  # For CORS
    allowed_referrers: list[str] = Field(default_factory=list)

    # Expiration
    expires_at: Optional[datetime] = None
    last_used_at: Optional[datetime] = None

    # Metadata
    created_at: datetime
    updated_at: Optional[datetime] = None
    created_by: str
    tags: list[str] = Field(default_factory=list)
    metadata: dict[str, Any] = Field(default_factory=dict)


class APIKeyCreate(BaseModel):
    """Create API key request"""
    name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    key_type: APIKeyType = APIKeyType.PRIVATE
    workspace_id: Optional[str] = None
    permissions: Optional[APIKeyPermissions] = None
    rate_limits: Optional[RateLimitConfig] = None
    allowed_ips: list[str] = Field(default_factory=list)
    allowed_domains: list[str] = Field(default_factory=list)
    expires_in_days: Optional[int] = Field(None, ge=1, le=365)
    tags: list[str] = Field(default_factory=list)


class APIKeyUpdate(BaseModel):
    """Update API key request"""
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    permissions: Optional[APIKeyPermissions] = None
    rate_limits: Optional[RateLimitConfig] = None
    allowed_ips: Optional[list[str]] = None
    allowed_domains: Optional[list[str]] = None
    status: Optional[APIKeyStatus] = None
    tags: Optional[list[str]] = None


class APIKeyResponse(BaseModel):
    """API key response (includes full key only on creation)"""
    api_key: APIKey
    key: Optional[str] = None  # Full key, only returned on creation


class APIKeyListResponse(BaseModel):
    """List of API keys"""
    keys: list[APIKey]
    total: int


# Usage Tracking Models

class UsageRecord(BaseModel):
    """Single usage record"""
    id: str
    api_key_id: str
    endpoint: str
    method: str
    status_code: int
    response_time_ms: int
    request_size_bytes: int
    response_size_bytes: int
    ip_address: Optional[str] = None
    user_agent: Optional[str] = None
    timestamp: datetime
    metadata: dict[str, Any] = Field(default_factory=dict)


class UsageStats(BaseModel):
    """Usage statistics for a period"""
    api_key_id: str
    period_start: datetime
    period_end: datetime

    # Request counts
    total_requests: int = 0
    successful_requests: int = 0
    failed_requests: int = 0
    rate_limited_requests: int = 0

    # By endpoint
    requests_by_endpoint: dict[str, int] = Field(default_factory=dict)
    requests_by_method: dict[str, int] = Field(default_factory=dict)
    requests_by_status: dict[int, int] = Field(default_factory=dict)

    # Performance
    avg_response_time_ms: float = 0
    p50_response_time_ms: float = 0
    p95_response_time_ms: float = 0
    p99_response_time_ms: float = 0

    # Data transfer
    total_request_bytes: int = 0
    total_response_bytes: int = 0

    # Unique
    unique_ips: int = 0
    unique_endpoints: int = 0


class UsageStatsResponse(BaseModel):
    """Usage stats response"""
    stats: UsageStats
    daily_breakdown: list[dict[str, Any]] = Field(default_factory=list)
    top_endpoints: list[dict[str, Any]] = Field(default_factory=list)


# Rate Limit Models

class RateLimitStatus(BaseModel):
    """Current rate limit status"""
    api_key_id: str
    period: RateLimitPeriod
    limit: int
    remaining: int
    reset_at: datetime
    is_limited: bool = False


class RateLimitResponse(BaseModel):
    """Rate limit info in response headers"""
    limit: int
    remaining: int
    reset: int  # Unix timestamp
    retry_after: Optional[int] = None  # Seconds until retry


# SDK Configuration Models

class SDKConfig(BaseModel):
    """SDK configuration for embedding"""
    api_key: str  # Public key
    base_url: str = "https://api.dataviz.bheemkodee.com"
    version: str = "1.0"

    # Feature flags
    enable_interactions: bool = True
    enable_export: bool = True
    enable_fullscreen: bool = True
    enable_comments: bool = False
    enable_refresh: bool = True

    # Theming
    theme: str = "light"  # light, dark, auto
    locale: str = "en"
    custom_css: Optional[str] = None

    # Security
    allowed_domains: list[str] = Field(default_factory=list)

    # Callbacks
    on_load: Optional[str] = None  # JS function name
    on_error: Optional[str] = None
    on_interaction: Optional[str] = None


class SDKInitOptions(BaseModel):
    """SDK initialization options"""
    container: str  # DOM selector or element ID
    resource_type: str  # dashboard, chart
    resource_id: str
    config: SDKConfig

    # Display options
    width: str = "100%"
    height: str = "600px"
    responsive: bool = True

    # Initial state
    filters: Optional[dict[str, Any]] = None
    variables: Optional[dict[str, Any]] = None


# Code Generation Models

class CodeSnippet(BaseModel):
    """Generated code snippet"""
    language: SDKLanguage
    code: str
    filename: Optional[str] = None
    description: Optional[str] = None
    dependencies: list[str] = Field(default_factory=list)


class CodeGenerationRequest(BaseModel):
    """Request to generate code"""
    language: SDKLanguage
    resource_type: str  # dashboard, chart, query
    resource_id: str
    include_auth: bool = True
    include_error_handling: bool = True
    include_types: bool = True  # For TS
    options: dict[str, Any] = Field(default_factory=dict)


class CodeGenerationResponse(BaseModel):
    """Generated code response"""
    snippets: list[CodeSnippet]
    setup_instructions: str
    documentation_url: Optional[str] = None


# API Documentation Models

class APIParameter(BaseModel):
    """API parameter documentation"""
    name: str
    type: str
    required: bool = False
    description: Optional[str] = None
    default: Optional[Any] = None
    enum_values: list[str] = Field(default_factory=list)
    example: Optional[Any] = None


class APIEndpoint(BaseModel):
    """API endpoint documentation"""
    path: str
    method: str
    summary: str
    description: Optional[str] = None
    category: EndpointCategory
    tags: list[str] = Field(default_factory=list)

    # Parameters
    path_params: list[APIParameter] = Field(default_factory=list)
    query_params: list[APIParameter] = Field(default_factory=list)
    body_schema: Optional[dict[str, Any]] = None

    # Response
    response_schema: Optional[dict[str, Any]] = None
    response_examples: dict[str, Any] = Field(default_factory=dict)

    # Security
    requires_auth: bool = True
    required_permissions: list[str] = Field(default_factory=list)
    rate_limited: bool = True

    # Metadata
    deprecated: bool = False
    version: str = "v1"


class APIEndpointGroup(BaseModel):
    """Group of related endpoints"""
    name: str
    description: Optional[str] = None
    category: EndpointCategory
    endpoints: list[APIEndpoint]


class APIDocumentation(BaseModel):
    """Full API documentation"""
    title: str = "Bheem DataViz API"
    version: str = "1.0.0"
    description: str = "REST API for Bheem DataViz platform"
    base_url: str = "https://api.dataviz.bheemkodee.com/api/v1"

    # Authentication
    auth_description: str = ""
    auth_examples: dict[str, str] = Field(default_factory=dict)

    # Endpoints
    groups: list[APIEndpointGroup] = Field(default_factory=list)

    # Schemas
    schemas: dict[str, Any] = Field(default_factory=dict)

    # SDKs
    sdk_languages: list[SDKLanguage] = Field(default_factory=list)
    sdk_download_urls: dict[str, str] = Field(default_factory=dict)


# Webhook Models (for SDK events)

class SDKWebhookEvent(BaseModel):
    """SDK event sent to webhooks"""
    id: str
    event_type: str  # embed.loaded, embed.error, embed.interaction
    api_key_id: str
    resource_type: str
    resource_id: str
    timestamp: datetime
    data: dict[str, Any] = Field(default_factory=dict)
    metadata: dict[str, Any] = Field(default_factory=dict)


class SDKWebhookConfig(BaseModel):
    """Webhook configuration for SDK events"""
    id: str
    api_key_id: str
    url: str
    secret: str  # For signature verification
    events: list[str] = Field(default_factory=list)  # Event types to send
    enabled: bool = True
    created_at: datetime
    updated_at: Optional[datetime] = None


# Constants

DEFAULT_RATE_LIMITS = RateLimitConfig(
    requests_per_minute=60,
    requests_per_hour=1000,
    requests_per_day=10000,
    burst_limit=10,
)

PUBLIC_KEY_RATE_LIMITS = RateLimitConfig(
    requests_per_minute=30,
    requests_per_hour=500,
    requests_per_day=5000,
    burst_limit=5,
)

SDK_LANGUAGES_INFO = {
    SDKLanguage.JAVASCRIPT: {
        "name": "JavaScript",
        "package": "@bheem/dataviz-sdk",
        "install": "npm install @bheem/dataviz-sdk",
    },
    SDKLanguage.TYPESCRIPT: {
        "name": "TypeScript",
        "package": "@bheem/dataviz-sdk",
        "install": "npm install @bheem/dataviz-sdk",
    },
    SDKLanguage.PYTHON: {
        "name": "Python",
        "package": "bheem-dataviz",
        "install": "pip install bheem-dataviz",
    },
    SDKLanguage.REACT: {
        "name": "React",
        "package": "@bheem/dataviz-react",
        "install": "npm install @bheem/dataviz-react",
    },
    SDKLanguage.VUE: {
        "name": "Vue",
        "package": "@bheem/dataviz-vue",
        "install": "npm install @bheem/dataviz-vue",
    },
    SDKLanguage.CURL: {
        "name": "cURL",
        "package": None,
        "install": None,
    },
}


# Helper Functions

def generate_api_key_prefix() -> str:
    """Generate API key prefix for identification"""
    import secrets
    return f"bv_{secrets.token_hex(4)}"


def hash_api_key(key: str) -> str:
    """Hash API key for storage"""
    import hashlib
    return hashlib.sha256(key.encode()).hexdigest()


def generate_full_api_key(prefix: str) -> str:
    """Generate full API key"""
    import secrets
    return f"{prefix}_{secrets.token_urlsafe(32)}"


def mask_api_key(key: str) -> str:
    """Mask API key for display"""
    if len(key) < 12:
        return "*" * len(key)
    return f"{key[:8]}...{key[-4:]}"


def validate_ip_address(ip: str) -> bool:
    """Validate IP address format"""
    import re
    ipv4_pattern = r'^(\d{1,3}\.){3}\d{1,3}$'
    ipv6_pattern = r'^([0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}$'
    cidr_pattern = r'^(\d{1,3}\.){3}\d{1,3}/\d{1,2}$'

    return bool(
        re.match(ipv4_pattern, ip) or
        re.match(ipv6_pattern, ip) or
        re.match(cidr_pattern, ip)
    )
